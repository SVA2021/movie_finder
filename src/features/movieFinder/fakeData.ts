import { TFullCard, TSmallCard } from "./movieFinderTypes";

// export const filmDataShort: TMovieShort = {
// 	filmId: 263531,
// 	nameRu: 'FakeNameRu',
// 	nameEn: 'FakeNameEng',
// 	type: 'FILM',
// 	year: '2000',
// 	description: 'FakeDescription',
// 	filmLength: 100,
// 	countries: [{ country: 'FakeCountry' }],
// 	genres: [{ genre: 'FakeGenre' }],
// 	rating: '85%',
// 	ratingVoteCount: 1000,
// 	posterUrl: 'https://kinopoiskapiunofficial.tech/images/posters/kp/301.jpg',
// 	posterUrlPreview: 'https://kinopoiskapiunofficial.tech/images/posters/kp_small/301.jpg',
// }

// export const fakeSmallCard: TSmallCard = {
// 	filmId: 263531,
// 	nameRu: 'FakeNameRu',
// 	nameEn: 'FakeNameEng',
// 	year: 2000,
// 	filmLength: 100,
// 	countries: [{ country: 'FakeCountry' }],
// 	genres: [{ genre: 'FakeGenre' }],
// 	rating: 7.9,
// 	ratingVoteCount: 1000,
// 	posterUrl: 'https://kinopoiskapiunofficial.tech/images/posters/kp/301.jpg',
// 	posterUrlPreview: 'https://kinopoiskapiunofficial.tech/images/posters/kp_small/301.jpg',
// }

// export const fakeSmallCard: TSmallCard = {
// 	"id": 395787,
// 	"nameRu": "Семь жизней",
// 	"nameEn": "Seven Pounds",
// 	"year": "2008",
// 	"filmLength": "01:58",
// 	"countries": [
// 		{
// 			"country": "США"
// 		}
// 	],
// 	"genres": [
// 		{
// 			"genre": "драма"
// 		},
// 		{
// 			"genre": "мелодрама"
// 		}
// 	],
// 	"rating": "8.2",
// 	"ratingVoteCount": 344640,
// 	"posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/395787.jpg",
// 	"posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/395787.jpg",
// 	"ratingChange": null
// }

export const fakeFullCard: TFullCard = {
	"kinopoiskId": 555,
	"imdbId": "tt0118715",
	"nameRu": "Большой Лебовски",
	"nameEn": null,
	"nameOriginal": "The Big Lebowski",
	"posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/555.jpg",
	"posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/555.jpg",
	"coverUrl": "https://avatars.mds.yandex.net/get-ott/239697/2a000001612cac1f659b8bf59d99a62d073c/orig",
	"logoUrl": null,
	"reviewsCount": 142,
	"ratingGoodReview": 93.3,
	"ratingGoodReviewVoteCount": 128,
	"ratingKinopoisk": 7.8,
	"ratingKinopoiskVoteCount": 242932,
	"ratingImdb": 8.1,
	"ratingImdbVoteCount": 792300,
	"ratingFilmCritics": 7.5,
	"ratingFilmCriticsVoteCount": 109,
	"ratingAwait": null,
	"ratingAwaitCount": 0,
	"ratingRfCritics": null,
	"ratingRfCriticsVoteCount": 2,
	"webUrl": "https://www.kinopoisk.ru/film/555/",
	"year": 1998,
	"filmLength": 117,
	"slogan": "It takes guys as simple as the Dude and Walter to make a story this complicated... and they'd really rather be bowling",
	"description": "Лос-Анджелес, 1991 год, война в Персидском заливе. Главный герой по прозвищу Чувак считает себя совершенно счастливым человеком. Его жизнь составляют игра в боулинг и выпивка. Но внезапно его счастье нарушается, гангстеры по ошибке принимают его за миллионера-однофамильца, требуют деньги, о которых он ничего не подозревает, и, ко всему прочему, похищают жену миллионера, будучи уверенными, что «муж» выплатит за нее любую сумму.",
	"shortDescription": "Тихого пофигиста по ошибке принимают за богача. Криминальная комедия братьев Коэн, которая разошлась на цитаты",
	"editorAnnotation": null,
	"isTicketsAvailable": false,
	"productionStatus": null,
	"type": "FILM",
	"ratingMpaa": "r",
	"ratingAgeLimits": "age18",
	"countries": [
		{
			"country": "США"
		},
		{
			"country": "Великобритания"
		}
	],
	"genres": [
		{
			"genre": "криминал"
		},
		{
			"genre": "комедия"
		}
	],
	"startYear": null,
	"endYear": null,
	"serial": false,
	"shortFilm": false,
	"completed": false,
	"hasImax": false,
	"has3D": false,
	"lastSync": "2022-06-05T12:58:15.228012"
}

export const fakeResponseByKeyword = {
  "total": 76,
  "totalPages": 4,
  "items": [
    {
      "kinopoiskId": 278522,
      "imdbId": "tt0903624",
      "nameRu": "Хоббит: Нежданное путешествие",
      "nameEn": null,
      "nameOriginal": "The Hobbit: An Unexpected Journey",
      "countries": [
        {
          "country": "США"
        },
        {
          "country": "Новая Зеландия"
        }
      ],
      "genres": [
        {
          "genre": "приключения"
        },
        {
          "genre": "боевик"
        },
        {
          "genre": "фэнтези"
        }
      ],
      "ratingKinopoisk": 8.1,
      "ratingImdb": 7.8,
      "year": 2012,
      "type": "FILM",
      "posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/278522.jpg",
      "posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/278522.jpg"
    },
  ]
};

export const TOP2 = {
  "pagesCount": 13,
  "films": [
    {
      "filmId": 395787,
      "nameRu": "Семь жизней",
      "nameEn": "Seven Pounds",
      "year": "2008",
      "filmLength": "01:58",
      "countries": [
        {
          "country": "США"
        }
      ],
      "genres": [
        {
          "genre": "драма"
        },
        {
          "genre": "мелодрама"
        }
      ],
      "rating": "8.2",
      "ratingVoteCount": 351371,
      "posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/395787.jpg",
      "posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/395787.jpg",
      "ratingChange": null
    },
    {
      "filmId": 458,
      "nameRu": "Корпорация монстров",
      "nameEn": "Monsters, Inc.",
      "year": "2001",
      "filmLength": "01:32",
      "countries": [
        {
          "country": "США"
        }
      ],
      "genres": [
        {
          "genre": "приключения"
        },
        {
          "genre": "фэнтези"
        },
        {
          "genre": "комедия"
        },
        {
          "genre": "мультфильм"
        },
        {
          "genre": "семейный"
        }
      ],
      "rating": "8.1",
      "ratingVoteCount": 409339,
      "posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/458.jpg",
      "posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/458.jpg",
      "ratingChange": null
    },
    {
      "filmId": 714248,
      "nameRu": "Песнь моря",
      "nameEn": "Song of the Sea",
      "year": "2014",
      "filmLength": "01:33",
      "countries": [
        {
          "country": "Франция"
        },
        {
          "country": "Дания"
        },
        {
          "country": "Ирландия"
        },
        {
          "country": "Бельгия"
        }
      ],
      "genres": [
        {
          "genre": "драма"
        },
        {
          "genre": "приключения"
        },
        {
          "genre": "фэнтези"
        },
        {
          "genre": "мультфильм"
        },
        {
          "genre": "семейный"
        }
      ],
      "rating": "8.1",
      "ratingVoteCount": 81444,
      "posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/714248.jpg",
      "posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/714248.jpg",
      "ratingChange": null
    },
    {
      "filmId": 530,
      "nameRu": "Игры разума",
      "nameEn": "A Beautiful Mind",
      "year": "2001",
      "filmLength": "02:15",
      "countries": [
        {
          "country": "США"
        }
      ],
      "genres": [
        {
          "genre": "драма"
        },
        {
          "genre": "мелодрама"
        },
        {
          "genre": "биография"
        }
      ],
      "rating": "8.5",
      "ratingVoteCount": 448759,
      "posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/530.jpg",
      "posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/530.jpg",
      "ratingChange": null
    },
    {
      "filmId": 327,
      "nameRu": "Крестный отец 2",
      "nameEn": "The Godfather: Part II",
      "year": "1974",
      "filmLength": "03:22",
      "countries": [
        {
          "country": "США"
        }
      ],
      "genres": [
        {
          "genre": "драма"
        },
        {
          "genre": "криминал"
        }
      ],
      "rating": "8.5",
      "ratingVoteCount": 182923,
      "posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/327.jpg",
      "posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/327.jpg",
      "ratingChange": null
    },
    {
      "filmId": 399,
      "nameRu": "Храброе сердце",
      "nameEn": "Braveheart",
      "year": "1995",
      "filmLength": "02:58",
      "countries": [
        {
          "country": "США"
        }
      ],
      "genres": [
        {
          "genre": "драма"
        },
        {
          "genre": "биография"
        },
        {
          "genre": "военный"
        },
        {
          "genre": "история"
        }
      ],
      "rating": "8.3",
      "ratingVoteCount": 253676,
      "posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/399.jpg",
      "posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/399.jpg",
      "ratingChange": null
    },
    {
      "filmId": 104938,
      "nameRu": "В погоне за счастьем",
      "nameEn": "The Pursuit of Happyness",
      "year": "2006",
      "filmLength": "01:57",
      "countries": [
        {
          "country": "США"
        }
      ],
      "genres": [
        {
          "genre": "драма"
        },
        {
          "genre": "биография"
        }
      ],
      "rating": "8.3",
      "ratingVoteCount": 313870,
      "posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/104938.jpg",
      "posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/104938.jpg",
      "ratingChange": null
    },
    {
      "filmId": 456,
      "nameRu": "Унесённые ветром",
      "nameEn": "Gone with the Wind",
      "year": "1939",
      "filmLength": "03:42",
      "countries": [
        {
          "country": "США"
        }
      ],
      "genres": [
        {
          "genre": "драма"
        },
        {
          "genre": "мелодрама"
        },
        {
          "genre": "военный"
        },
        {
          "genre": "история"
        }
      ],
      "rating": "8.4",
      "ratingVoteCount": 172328,
      "posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/456.jpg",
      "posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/456.jpg",
      "ratingChange": null
    },
    {
      "filmId": 77269,
      "nameRu": "Шерлок Холмс и доктор Ватсон: Знакомство",
      "nameEn": null,
      "year": "1979",
      "filmLength": "01:08",
      "countries": [
        {
          "country": "СССР"
        }
      ],
      "genres": [
        {
          "genre": "криминал"
        },
        {
          "genre": "детектив"
        }
      ],
      "rating": "8.6",
      "ratingVoteCount": 144506,
      "posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/77269.jpg",
      "posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/77269.jpg",
      "ratingChange": null
    },
    {
      "filmId": 450213,
      "nameRu": "Невидимая сторона",
      "nameEn": "The Blind Side",
      "year": "2009",
      "filmLength": "02:09",
      "countries": [
        {
          "country": "США"
        }
      ],
      "genres": [
        {
          "genre": "драма"
        },
        {
          "genre": "биография"
        },
        {
          "genre": "спорт"
        }
      ],
      "rating": "8.2",
      "ratingVoteCount": 231733,
      "posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/450213.jpg",
      "posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/450213.jpg",
      "ratingChange": null
    },
    {
      "filmId": 280172,
      "nameRu": "Как приручить дракона",
      "nameEn": "How to Train Your Dragon",
      "year": "2010",
      "filmLength": "01:38",
      "countries": [
        {
          "country": "США"
        }
      ],
      "genres": [
        {
          "genre": "приключения"
        },
        {
          "genre": "фэнтези"
        },
        {
          "genre": "комедия"
        },
        {
          "genre": "мультфильм"
        },
        {
          "genre": "семейный"
        }
      ],
      "rating": "8.2",
      "ratingVoteCount": 492918,
      "posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/280172.jpg",
      "posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/280172.jpg",
      "ratingChange": null
    },
    {
      "filmId": 7103,
      "nameRu": "Назад в будущее 3",
      "nameEn": "Back to the Future Part III",
      "year": "1990",
      "filmLength": "01:58",
      "countries": [
        {
          "country": "США"
        }
      ],
      "genres": [
        {
          "genre": "фантастика"
        },
        {
          "genre": "приключения"
        },
        {
          "genre": "вестерн"
        },
        {
          "genre": "комедия"
        }
      ],
      "rating": "8.2",
      "ratingVoteCount": 310734,
      "posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/7103.jpg",
      "posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/7103.jpg",
      "ratingChange": null
    },
    {
      "filmId": 42172,
      "nameRu": "Тайна третьей планеты",
      "nameEn": null,
      "year": "1981",
      "filmLength": "00:50",
      "countries": [
        {
          "country": "СССР"
        }
      ],
      "genres": [
        {
          "genre": "фантастика"
        },
        {
          "genre": "приключения"
        },
        {
          "genre": "мультфильм"
        },
        {
          "genre": "семейный"
        }
      ],
      "rating": "8.1",
      "ratingVoteCount": 119051,
      "posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/42172.jpg",
      "posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/42172.jpg",
      "ratingChange": null
    },
    {
      "filmId": 9691,
      "nameRu": "Бесславные ублюдки",
      "nameEn": "Inglourious Basterds",
      "year": "2009",
      "filmLength": "02:33",
      "countries": [
        {
          "country": "США"
        },
        {
          "country": "Германия"
        }
      ],
      "genres": [
        {
          "genre": "драма"
        },
        {
          "genre": "боевик"
        },
        {
          "genre": "комедия"
        },
        {
          "genre": "военный"
        }
      ],
      "rating": "7.9",
      "ratingVoteCount": 424593,
      "posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/9691.jpg",
      "posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/9691.jpg",
      "ratingChange": null
    },
    {
      "filmId": 5277,
      "nameRu": "Мулан",
      "nameEn": "Mulan",
      "year": "1998",
      "filmLength": "01:24",
      "countries": [
        {
          "country": "США"
        }
      ],
      "genres": [
        {
          "genre": "приключения"
        },
        {
          "genre": "фэнтези"
        },
        {
          "genre": "комедия"
        },
        {
          "genre": "военный"
        },
        {
          "genre": "мультфильм"
        },
        {
          "genre": "семейный"
        },
        {
          "genre": "мюзикл"
        }
      ],
      "rating": "8.1",
      "ratingVoteCount": 242690,
      "posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/5277.jpg",
      "posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/5277.jpg",
      "ratingChange": null
    },
    {
      "filmId": 278522,
      "nameRu": "Хоббит: Нежданное путешествие",
      "nameEn": "The Hobbit: An Unexpected Journey",
      "year": "2012",
      "filmLength": "02:49",
      "countries": [
        {
          "country": "США"
        },
        {
          "country": "Новая Зеландия"
        }
      ],
      "genres": [
        {
          "genre": "приключения"
        },
        {
          "genre": "боевик"
        },
        {
          "genre": "фэнтези"
        }
      ],
      "rating": "8.1",
      "ratingVoteCount": 455782,
      "posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/278522.jpg",
      "posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/278522.jpg",
      "ratingChange": null
    },
    {
      "filmId": 5273,
      "nameRu": "Шрэк 2",
      "nameEn": "Shrek 2",
      "year": "2004",
      "filmLength": "01:33",
      "countries": [
        {
          "country": "США"
        }
      ],
      "genres": [
        {
          "genre": "мелодрама"
        },
        {
          "genre": "приключения"
        },
        {
          "genre": "фэнтези"
        },
        {
          "genre": "комедия"
        },
        {
          "genre": "мультфильм"
        },
        {
          "genre": "семейный"
        }
      ],
      "rating": "7.7",
      "ratingVoteCount": 257786,
      "posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/5273.jpg",
      "posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/5273.jpg",
      "ratingChange": null
    },
    {
      "filmId": 441,
      "nameRu": "Принцесса Мононоке",
      "nameEn": "Mononoke-hime",
      "year": "1997",
      "filmLength": "02:14",
      "countries": [
        {
          "country": "Япония"
        }
      ],
      "genres": [
        {
          "genre": "драма"
        },
        {
          "genre": "приключения"
        },
        {
          "genre": "фэнтези"
        },
        {
          "genre": "мультфильм"
        },
        {
          "genre": "аниме"
        }
      ],
      "rating": "8.2",
      "ratingVoteCount": 137308,
      "posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/441.jpg",
      "posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/441.jpg",
      "ratingChange": null
    },
    {
      "filmId": 102124,
      "nameRu": "Пираты Карибского моря: На краю света",
      "nameEn": "Pirates of the Caribbean: At World's End",
      "year": "2007",
      "filmLength": "02:49",
      "countries": [
        {
          "country": "США"
        }
      ],
      "genres": [
        {
          "genre": "приключения"
        },
        {
          "genre": "боевик"
        },
        {
          "genre": "фэнтези"
        }
      ],
      "rating": "8.0",
      "ratingVoteCount": 440467,
      "posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/102124.jpg",
      "posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/102124.jpg",
      "ratingChange": null
    },
    {
      "filmId": 470191,
      "nameRu": "Невероятный мир глазами Энцо",
      "nameEn": "The Art of Racing in the Rain",
      "year": "2019",
      "filmLength": "01:49",
      "countries": [
        {
          "country": "США"
        }
      ],
      "genres": [
        {
          "genre": "драма"
        },
        {
          "genre": "мелодрама"
        },
        {
          "genre": "комедия"
        },
        {
          "genre": "семейный"
        },
        {
          "genre": "спорт"
        }
      ],
      "rating": "8.2",
      "ratingVoteCount": 94545,
      "posterUrl": "https://kinopoiskapiunofficial.tech/images/posters/kp/470191.jpg",
      "posterUrlPreview": "https://kinopoiskapiunofficial.tech/images/posters/kp_small/470191.jpg",
      "ratingChange": null
    }
  ]
};