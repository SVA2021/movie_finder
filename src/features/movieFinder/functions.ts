export const normalizeDateString = (date: number): string => {
	let result;
	(date < 10) ? result = '0' + String(date) : result = String(date);
	return result;
}

export const convertNumToTime = (num: number | null): string => {
	let result;
	if (num === null) return '';
	let hours = Math.floor(num % 60);
	let minutes = num % 60;

	result = (hours === 0) ? `${minutes}мин ` : `${hours}ч ${minutes}мин`;
	return result;
}

export const convertRatingToColor = (num: number| null): string => {
	if (num === null) return '';
	if (num <= 0 || num > 10) return 'white';
	if (num >= 8) return 'green';
	if (num >= 6) return 'orange';
	return 'red';
}