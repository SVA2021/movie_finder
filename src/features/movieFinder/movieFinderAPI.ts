import axios from "axios";
import { API_OFFICIAL } from "../../private/key";
import { TTopList } from "./movieFinderTypes";

export const BASE_URL = 'https://kinopoiskapiunofficial.tech/';
const FILMS = 'api/v2.2/films/';
const TOP = 'top/';

const movieFinderInstance = axios.create({
	baseURL: BASE_URL,
	headers: {
		'X-API-KEY': API_OFFICIAL,
		'Content-Type': 'application/json',
	},
});

export const movieFinderAPI = {

	getTopList(topList: TTopList = 'TOP_250_BEST_FILMS', page: number = 1) {
		let query = `${FILMS}${TOP}?page=${page}`;
		return movieFinderInstance.get(query)
			.then(response => {
				console.log(response);
				return response;
			})
			.catch(e => {
				console.log(e);
				return e;
			});
	},

	getMovieData(id: number = 1) {
		let query = `${FILMS}${id}`;
		return movieFinderInstance.get(query)
			.then(response => {
				console.log(response);
				return response;
			})
			.catch(e => {
				console.log(e);
				return e;
			});
	},

	getMoviesByKeyword(keyword: string | undefined, page: number = 1) {
		let query = `${FILMS}?keyword=${keyword}&page=${page}`;
		return movieFinderInstance.get(query)
			.then(response => {
				console.log(response);
				return response;
			})
			.catch(e => {
				console.log(e);
				return e;
			});
	},

	getSeries(page: number = 1) {
		let query = `${FILMS}?type=TV_SERIES&page=${page}`;
		return movieFinderInstance.get(query)
			.then(response => {
				console.log(response);
				return response;
			})
			.catch(e => {
				console.log(e);
				return e;
			});
	},
}
