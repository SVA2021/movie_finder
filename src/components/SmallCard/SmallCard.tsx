import {Card, CardActionArea, CardContent, CardMedia, Typography} from "@mui/material";
import {FC} from 'react';
import {useNavigate} from "react-router-dom";
import {RatingBadge} from "../../components";
import {TSmallCard} from "../../features/movieFinder/movieFinderTypes";
import {normalizeTimeString} from "../../utils";

interface TSmallCardProps {
	item: TSmallCard
}

export const SmallCard: FC<TSmallCardProps> = ({item}) => {

	const {id, nameRu, nameEn, year, filmLength, rating, posterUrlPreview} = item;
	const navigate = useNavigate();
	const time = normalizeTimeString(filmLength);

	function getDetails() {
		navigate(`details/${id}`);
	}

	return (
		<Card data-testid={'smallCard'} sx={{position: 'relative', }}>
			<CardActionArea onClick={() => getDetails()} sx={{position: 'relative', height: '100%'}}>
				<CardMedia
					component="img"
					width="100%"
					image={posterUrlPreview}
					alt={nameRu?.toString() || nameEn?.toString() || ''}
				/>
				<CardContent>
					<Typography gutterBottom variant="body1" component="p">{nameRu}</Typography>
					{nameEn &&
						<Typography fontStyle={'italic'} gutterBottom variant="body2" component="p">
							{nameEn}
						</Typography>}
					<Typography variant="body1" component="p">
						{year},  {time}
					</Typography>
				</CardContent>
			</CardActionArea>
			<RatingBadge rating={rating} position={'left'} />
		</Card>
	)
};