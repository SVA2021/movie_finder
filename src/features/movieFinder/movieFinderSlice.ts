import {AnyAction, createAsyncThunk, createSlice, PayloadAction} from '@reduxjs/toolkit';
import {RootState} from '../../app/store';
import {movieFinderAPI} from './movieFinderAPI';
import {
	TMovieDetails,
	TFullCard, THomePageCurrent, TSearchResponseData,
	TStatus, TTopData, TTopList, TTopResponse, TUser
} from "./movieFinderTypes";

interface TMovieFinderState {
	isAuth: boolean
	user: TUser | null
	status: TStatus
	detailsPage: TMovieDetails
	fullCard: TFullCard | null
	fullCardExtraInfo: null
	homePage: {[key in TTopList]: TTopData | null}
	movies: TSearchResponseData | null
	series: TSearchResponseData | null
	searchResult: TSearchResponseData
	error: string | null
}

export const initialState: TMovieFinderState = {
	isAuth: false,
	user: null,
	status: 'idle',
	detailsPage: {
		movie: null,
	},
	fullCard: null,
	fullCardExtraInfo: null,
	homePage: {
		TOP_250_BEST_FILMS: null,
		TOP_100_POPULAR_FILMS: null,
		TOP_AWAIT_FILMS: null,
	},
	movies: null,
	series: null,
	searchResult: {
		total: null,
		totalPages: null,
		currentPage: 1,
		keyword: null,
		items: {},
	},
	error: null
};

export const getTopListAsync = createAsyncThunk<TTopResponse, THomePageCurrent, {rejectValue: string}>(
	'movieFinder/getTopList',
	async function ({type, page}: THomePageCurrent, {rejectWithValue}) {
		const response = await movieFinderAPI.getTopList(type, page);
		if (response.status === 200) {
			return {...response.data, type, page, }
		} else {
			return rejectWithValue('Server Error!');
		}
	}
);

export const getMovieDataAsync = createAsyncThunk<TFullCard, number, {rejectValue: string}>(
	'movieFinder/getMovieData',
	async function (id: number, {rejectWithValue}) {
		const response = await movieFinderAPI.getMovieData(id);
		if (response.status === 200) {
			return response.data
		} else {
			return rejectWithValue('Server Error!');
		}
	}
);

export const movieFinderSlice = createSlice({
	name: 'movieFinder',
	initialState,

	reducers: {
		setIsAuth: (state, action: PayloadAction<boolean>) => {
			state.isAuth = action.payload;
		},
		setFullCard: (state, action: PayloadAction<null | TFullCard>) => {
			state.fullCard = action.payload;
		},
	},
	extraReducers: (builder) => {
		builder
			.addCase(getTopListAsync.pending, (state) => {
				state.status = 'loading';
			})
			.addCase(getTopListAsync.fulfilled, (state, action: PayloadAction<TTopResponse>) => {
				state.status = 'idle';
				state.homePage[action.payload.type] = {
					page: action.payload.page,
					pagesCount: action.payload.pagesCount,
					data: (state.homePage[action.payload.type] === null)
						? action.payload.films.map((v) => ({...v, id: v.filmId}))
						: [...state.homePage[action.payload.type]!.data, ...action.payload.films.map((v) => ({...v, id: v.filmId}))],
				}
			})
			.addCase(getMovieDataAsync.pending, (state) => {
				state.status = 'loading';
			})
			.addCase(getMovieDataAsync.fulfilled, (state, action: PayloadAction<TFullCard>) => {
				state.status = 'idle';
				state.detailsPage.movie = action.payload;
			})
			.addMatcher(isError, (state, action: PayloadAction<string>) => {
				state.error = action.payload;
				state.status = 'failed';
			});
	},
});

export const {setIsAuth, setFullCard, } = movieFinderSlice.actions;

export const selectIsAuth = (state: RootState) => state.movieFinder.isAuth;
export const selectStatus = (state: RootState) => state.movieFinder.status;
export const selectHomePage = (state: RootState) => state.movieFinder.homePage;
export const selectDetails = (state: RootState) => state.movieFinder.detailsPage.movie;
export const selectSearchResult = (state: RootState) => state.movieFinder.searchResult;

export default movieFinderSlice.reducer;

function isError(action: AnyAction) {
	return action.type.endsWith('rejected');
}