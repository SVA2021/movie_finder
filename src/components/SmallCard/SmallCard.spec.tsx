import { fireEvent, render, screen } from '@testing-library/react';
import { filmDataShort } from '../../features/movieFinder/fakeData';
import SmallCard from './SmallCard';

// const filmDataShort = {
// 	filmId: 100,
// 	nameRu: 'FakeNameRu',
// 	nameEng: 'FakeNameEng',
// 	nameOriginal: 'OriginalFakeName',
// 	posterUrl: 'https://kinopoiskapiunofficial.tech/images/posters/kp/301.jpg',
// 	posterUrlPreview: 'https://avatars.mds.yandex.net/get-kinopoisk-image/1704946/eed1de3a-5400-43b3-839e-22490389bf54/360',
// 	relationType: 'SIMILAR',
// }

describe('render SmallCard', () => {

	const handleClicker = jest.fn();
	const setUp = () => render(<SmallCard {...filmDataShort} handleClick={ handleClicker()} />);

	it('snapshot test', () => {
		setUp();
		expect(screen.getByTestId('smallCard')).toMatchSnapshot();
	})
	it('find movie name RU', () => {
		setUp();
		expect(screen.getByText(/FakeNameRu/i)).toBeInTheDocument();
	})
	it('find movie name Eng', () => {
		setUp();
		expect(screen.getByText(/FakeNameEng/i)).toBeInTheDocument();
	})
	it('img uses correct src', () => {
		setUp();
		const image = screen.getByAltText('FakeNameEng');
		expect(image).toHaveAttribute('src', 'https://avatars.mds.yandex.net/get-kinopoisk-image/1704946/eed1de3a-5400-43b3-839e-22490389bf54/360');
	});
	it('click on Card', () => {
		setUp();
		fireEvent.click(screen.getByTestId('smallCard'));
		expect(handleClicker).toHaveBeenCalledTimes(1);
	});

	screen.debug();
})