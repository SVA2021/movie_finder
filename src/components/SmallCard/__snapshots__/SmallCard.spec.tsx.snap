// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render SmallCard snapshot test 1`] = `
<div
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
  data-testid="smallCard"
>
  <button
    class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
    tabindex="0"
    type="button"
  >
    <img
      alt="FakeDescription"
      class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
      src="https://kinopoiskapiunofficial.tech/images/posters/kp_small/301.jpg"
      width="100%"
    />
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
      >
        263531
        , 
        FakeNameRu
        , 
        FakeNameEng
        , 
        100
        , 
        https://kinopoiskapiunofficial.tech/images/posters/kp_small/301.jpg
      </div>
      <p
        class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
      >
        Lizards are a widespread group of squamate reptiles, with over 6,000 species, ranging across all continents except Antarctica
      </p>
    </div>
    <span
      class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
    />
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
